### PR TITLE
add MatrixUnionRows IR node

### PR DESCRIPTION
**Pull request by catoverdrive**
keeps the behavior of MatrixTable.unionRows for now; this PR just moves it into IR.

### CODE REVIEW

File: src/main/scala/is/hail/expr/ir/Simplify.scala [hunk 1]
```diff
@@ -191,6 +191,12 @@ object Simplify {
           case c => Some(c)
         })
 
+      case MatrixUnionRows(children) if children.exists(_.isInstanceOf[MatrixUnionRows]) =>
+        MatrixUnionRows(children.flatMap {
+          case u: MatrixUnionRows => u.children
+          case c => Some(c)
+        })
+
```
Comment (tpoterba) [inline src/main/scala/is/hail/expr/ir/Simplify.scala]: add two to invert MatrixRowsTable(MatrixUnionRows()) and same for MatrixColsTable?

File: src/main/scala/is/hail/expr/ir/PruneDeadFields.scala [hunk 2]
```diff
@@ -841,5 +851,31 @@ object PruneDeadFields {
...
+    upcastCols: Boolean = true,
+    upcastGlobals: Boolean = true,
+    upcastEntries: Boolean = true): MatrixIR = {
+
+    if (ir.typ == rType || !(upcastRows || upcastCols || upcastGlobals || upcastEntries))
+      ir
+    else {
+      var mt = ir
+      if (upcastEntries && mt.typ.entryType != rType.entryType)
+        mt = MatrixMapEntries(mt, SelectFields(Ref("g", mt.typ.entryType), rType.entryType.fieldNames))
```
Comment (tpoterba) [inline src/main/scala/is/hail/expr/ir/PruneDeadFields.scala]: this won't work -- need to support the full nested type. Just use the existing `upcast` for value IRs.

File: src/main/scala/is/hail/expr/ir/PruneDeadFields.scala [hunk 3]
```diff
@@ -707,6 +710,13 @@ object PruneDeadFields {
           colFields.filter(childFieldSet.contains),
           partitionKey,
           nPartitions)
+      case MatrixUnionRows(children) =>
```
Comment (tpoterba) [inline src/main/scala/is/hail/expr/ir/PruneDeadFields.scala]: can you fix TableUnion too?

Comment (catoverdrive) [reply src/main/scala/is/hail/expr/ir/PruneDeadFields.scala]: yeah--fixing in a separate PR.

File: src/main/scala/is/hail/expr/ir/PruneDeadFields.scala (hunk 2 above)
Comment (catoverdrive) [reply src/main/scala/is/hail/expr/ir/PruneDeadFields.scala]: ah, good call.

**Pull request merged**